### PR TITLE
GetProcessInfo3 diagnostic command

### LINKED ISSUE #74476: Add ProcessInfo3 diagnostic IPC command to include libc type or equivalent
For diagnostic tools that want to load an ICorProfilerCallback implementation using the diagnostic communication channel's [AttachProfiler](https://github.com/dotnet/diagnostics/blob/main/documentation/design-docs/ipc-protocol.md#attachprofiler) command need to understand the architecture and libc type (glibc vs musl libc) in order to provide a path to profiler assembly with the matching architecture and libc type.

For tools that are not running in the same process namespace (e.g. dotnet-monitor), this is nearly impossible to determine based on any of the information available while the target process is suspended. The latest process info ([ProcessInfo2](https://github.com/dotnet/diagnostics/blob/main/documentation/design-docs/ipc-protocol.md#processinfo2)) command has OS and architecture information, but lacks libc type information. The environment block doesn't contain enough information to form a heuristic either; the following is an example environment block of an ASP.NET 6 Alpine container:

```json
{
  "PATH":"/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
  "HOSTNAME":"deploy-exampleapp-75cc844674-62w2q",
  "ASPNETCORE_URLS":"http://+:8080",
  "DOTNET_RUNNING_IN_CONTAINER":"true",
  "DOTNET_SYSTEM_GLOBALIZATION_INVARIANT":"true",
  "DOTNET_VERSION":"6.0.8",
  "ASPNET_VERSION":"6.0.8",
  "DOTNET_DiagnosticPorts":"/diag/port",
  "KUBERNETES_PORT_443_TCP_PORT":"443",
  "KUBERNETES_PORT_443_TCP_ADDR":"10.0.0.1",
  "KUBERNETES_SERVICE_HOST":"10.0.0.1",
  "KUBERNETES_SERVICE_PORT":"443",
  "KUBERNETES_SERVICE_PORT_HTTPS":"443",
  "KUBERNETES_PORT":"tcp://10.0.0.1:443",
  "KUBERNETES_PORT_443_TCP":"tcp://10.0.0.1:443",
  "KUBERNETES_PORT_443_TCP_PROTO":"tcp",
  "HOME":"/root"
}
```

The diagnostic pipe protocol should be updated with a new command (`ProcessInfo3`) that contains enough information to differentiate the libc type. Alternatively, this command could return the non-distribution specific runtime identifier (e.g. `linux-x64`, `linux-musl-x64`, `osx-arm64`, `win-x86`, etc), which is commonly used for packing assets for platform variations into the same package.

### PR DESCRIPTION
**Pull request by jander-msft**
Add the GetProcessInfo3 command to the diagnostic server. This allows for retrieving the portable runtime identifier which informs diagnostic tools which variants of native libraries may be loaded into the target process. The portable runtime identifier is statically calculated at compile time based on the build target information for the libraries that link in the diagnostic server.

Corresponding diagnostics change: https://github.com/dotnet/diagnostics/pull/3985

closes #74476

### CODE REVIEW

File: src/native/eventpipe/ds-sources.c [hunk 1]
```diff
@@ -16,6 +16,52 @@
...
+#else
+#define PORTABLE_RID_OS "unix"
+#endif
+
+#elif defined(TARGET_WINDOWS)
+#define PORTABLE_RID_OS "win"
+#endif
+
+#if defined(PORTABLE_RID_OS) && defined(PORTABLE_RID_ARCH)
+const ep_char8_t* _ds_portable_rid_info = PORTABLE_RID_OS "-" PORTABLE_RID_ARCH;
```
Comment (jander-msft) [inline src/native/eventpipe/ds-sources.c:60]: I would like to ask if this is actually composing a "portable RID" if it includes the architecture portion. If not, then there's two ways to resolve this:
- Rename it to what it actually represents (non-distro specific RID) but with a better concise descriptive name
- Remove the architecture portion and only report the OS part, namely `linux-bionic`, `linux-musl`, `linux`, `osx`, `win`, and `unix`. The architecture can already be retrieved from the architecture field on the process info structures.

Comment (jander-msft) [reply src/native/eventpipe/ds-sources.c:60]: Host policy seems to indicate that including architecture allowed for portable RIDs:
- https://github.com/dotnet/runtime/blob/main/src/native/corehost/hostpolicy/deps_format.cpp#L268
- https://github.com/dotnet/runtime/blob/main/src/native/corehost/hostpolicy/deps_format.cpp#L190

File: src/native/eventpipe/ds-sources.c [hunk 2]
```diff
@@ -16,6 +16,52 @@
...
+#else
+const ep_char8_t* _ds_portable_rid_info = "unknown";
```
Comment (jkotas) [inline src/native/eventpipe/ds-sources.c:62]: I think it would be better to get a build error if the ifdefs above do not cover the OS or architecture.

Comment (jander-msft) [reply src/native/eventpipe/ds-sources.c]: Fixed

**Comment by jander-msft** [comment]
Could I get a new build ("/azp run") for the current set of changes? Seems some Azure service hiccupped while this one was starting.

**Comment by jander-msft** [comment]
With the prior change, all jobs and tests were green except one where the new test was erroneously included. That was fixed in the latest commit. With that, I think this is ready for review.

File: src/native/eventpipe/ds-process-protocol.c [hunk 3]
```diff
@@ -352,6 +358,166 @@ ds_process_info_2_payload_fini (DiagnosticsProcessInfo2Payload *payload)
...
+	size += sizeof(uint32_t);
+	size += (payload->managed_entrypoint_assembly_name != NULL) ?
+		(ep_rt_utf16_string_len (payload->managed_entrypoint_assembly_name) + 1) * sizeof(ep_char16_t) : 0;
+
+	size += sizeof(uint32_t);
+	size += (payload->clr_product_version != NULL) ?
+		(ep_rt_utf16_string_len (payload->clr_product_version) + 1) * sizeof(ep_char16_t) : 0;
+
+	size += sizeof(uint32_t);
+	size += (payload->clr_product_version != NULL) ?
```
Comment (jander-msft) [inline src/native/eventpipe/ds-process-protocol.c:410]:
```suggestion
	size += (payload->portable_rid != NULL) ?
```

**Comment by jander-msft** [comment]
@tommcdon could I get someone from diagnostics to review this?

**Review by davmason — APPROVED** [review]
LGTM, but I want to give @lateralusX a chance to look before mergin

**Comment by lateralusX** [comment]
Sorry, was out on vacation but changes LGTM! One thought I had is that the portable rid will not be as specific as the OS where Mono will give you different identifiers for `iOS/tvOS` and `Android`, the portable rid will place all of them under each platforms inherited rid, so `iOS/tvOS` will present itself as `unix `and `Android` as `linux-bionic`, but I guess is fine, and it is always possible to take a look at the OS parameter to get more details around the specific OS. One platform that might not give you the right portable rid with current change is `maccatalyst` that should present itself similar to `iOS` (https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.NETCore.Platforms/src/runtime.json#L2296), but will with current change present itself as `osx`, so that probably needs to be adjusted. Having that said, `maccatalyst` is not currently included as an OS either, so when adjusting that we should probably also adjust that in https://github.com/dotnet/runtime/blob/main/src/native/eventpipe/ep-event-source.c#L15.

**Comment by jander-msft** [comment]
The intent of this portable RID is to provide a mechanism to differentiate the C library types (e.g. gnu, musl, bionic for Linux) for the target without having to know the specifics of the target so that native libraries with a portable notion could be loaded and executed in the target process; portable meaning not taking a hard dependency on specific distribution or version. We could have reported the literal C library type instead, but then we'd have to make up monikers or completely ignore other targets. Additionally, there seems to be support via library systems (e.g. NuGet) for providing libraries using this generic portable RID notion.

That being said, if something that is targeting a maccatalyst target needs to build a native library that is more specific than `iOS-x64` or `iOS-arm64` in order for it to be loaded and excitable, then these specifically can be carved out but should be done so in the most generic terms. For example, it seems that Void Linux supports both musl and gnu libc, but by default has gnu libc; it's RID in this scheme should not be `void-x64` but should be `linux-x64` or `linux-musl-x64`.

**Pull request merged**